minio/credentials/providers.py

```diff
@@ -17,70 +17,181 @@
 """Credential providers."""
 
 import configparser
+import ipaddress
 import json
 import os
+import socket
 import sys
+import time
+from abc import ABCMeta, abstractmethod
 from datetime import datetime, timedelta
+from urllib.parse import urlencode
+from xml.etree import ElementTree
 
 import urllib3
-from urllib3.exceptions import HTTPError, ResponseError
 
-from .credentials import Provider, Value
+from minio.helpers import get_sha256_hexdigest
+from minio.signer import sign_v4
+
+from .credentials import Credentials
 
 RFC3339NANO = "%Y-%m-%dT%H:%M:%S.%fZ"
 RFC3339 = "%Y-%m-%dT%H:%M:%SZ"
+_MIN_DURATION_SECONDS = timedelta(minutes=15).total_seconds()
+_MAX_DURATION_SECONDS = timedelta(days=7).total_seconds()
+_DEFAULT_DURATION_SECONDS = timedelta(hours=1).total_seconds()
+_XML_NS = {
+    "s3": "http://s3.amazonaws.com/doc/2006-03-01/",
+    "sts": "https://sts.amazonaws.com/doc/2011-06-15/",
+}
+
+
+def _parse_credentials(data, result_path):
+    """Parse data containing credentials XML."""
+
+    root = ElementTree.fromstring(data)
+    credentials = root.find("sts:" + result_path, _XML_NS).find(
+        "sts:Credentials", _XML_NS)
+
+    access_key = credentials.find("sts:AccessKeyId", _XML_NS).text
+    secret_key = credentials.find("sts:SecretAccessKey", _XML_NS).text
+    session_token = credentials.find("sts:SessionToken", _XML_NS).text
+    text = credentials.find("sts:Expiration", _XML_NS).text
+    try:
+        expiration = datetime.strptime(text, RFC3339NANO)
+    except ValueError:
+        expiration = datetime.strptime(text, RFC3339)
+
+    return Credentials(access_key, secret_key, session_token, expiration)
+
+
+def _urlopen(http_client, method, url, body=None, headers=None):
+    """Wrapper of urlopen() handles HTTP status code."""
+    res = http_client.urlopen(method, url, body=body, headers=headers)
+    if res.status not in [200, 204, 206]:
+        raise ValueError(
+            "{0} failed with HTTP status code {1}".format(url, res.status),
+        )
+    return res
+
+
+class Provider:  # pylint: disable=too-few-public-methods
+    """Credential retriever."""
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def retrieve(self):
+        """Retrieve credentials and its expiry if available."""
 
 
 class AssumeRoleProvider(Provider):
     """Assume-role credential provider."""
 
-    def __init__(self, provider_func):
-        self._provider_func = provider_func
-        self._value = None
-        self._expiry = None
+    def __init__(
+            self, sts_endpoint, access_key, secret_key, duration_seconds=0,
+            policy=None, region=None, role_arn=None, role_session_name=None,
+            external_id=None, http_client=None,
+    ):
+        self._sts_endpoint = sts_endpoint
+        self._access_key = access_key
+        self._secret_key = secret_key
+        self._region = region
```

Comment:
shouldn't we fallback to "us-east-1" as default if `region` is not specified?

Reply:
I see `region` is optional in `minio-go`; however for signature v4 region must be needed. I will check why `region` is optional in `minio-go` and fix accordingly here.

Reply:
I got clarified on this. We would need to use empty string as region if not provided. Fixing accordingly.